lnrpc/walletrpc/walletkit_server.go

```diff
@@ -542,3 +548,57 @@ func (w *WalletKit) BumpFee(ctx context.Context,
 
 	return &BumpFeeResponse{}, nil
 }
+
+// DeriveSharedKey returns a shared secret key by performing Diffie-Hellman key
+// derivation between the ephemeral public key in the request and the node's
+// identity public key. The resulting shared public key is hashed with sha256
+// and the final key length therefore is 256bit.
+func (w *WalletKit) DeriveSharedKey(ctx context.Context, in *SharedKeyRequest) (
+	*SharedKeyResponse, error) {
+
+	if len(in.EphemeralPubkey) != 33 {
+		return nil, fmt.Errorf("ephemeral pubkey must be " +
+			"serialized in compressed format")
+	}
+	ephemeralPubkey, err := btcec.ParsePubKey(
+		in.EphemeralPubkey, btcec.S256(),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse pubkey: %v", err)
+	}
+
+	// Derive our node's private key from the key ring.
+	secretKeyRing, ok := w.cfg.KeyRing.(keychain.SecretKeyRing)
+	if !ok {
+		return nil, fmt.Errorf("secret key ring not available")
+	}
+	idPrivKey, err := secretKeyRing.DerivePrivKey(keychain.KeyDescriptor{
+		KeyLocator: keychain.KeyLocator{
+			Family: keychain.KeyFamilyNodeKey,
+			Index:  0,
+		},
+	})
+	if err != nil {
+		err := fmt.Errorf("unable to derive node private key: %v", err)
+		log.Error(err)
+		return nil, err
+	}
+	idPrivKey.Curve = btcec.S256()
+
+	// Derive the shared key using ECDH and hashing the serialized
+	// compressed shared point.
+	sharedKeyHash := ecdh(ephemeralPubkey, idPrivKey)
+	return &SharedKeyResponse{SharedKey: sharedKeyHash}, nil
+}
+
+// ecdh performs an ECDH operation between pub and priv. The returned value is
+// the sha256 of the compressed shared point.
+func ecdh(pub *btcec.PublicKey, priv *btcec.PrivateKey) []byte {
+	s := &btcec.PublicKey{}
+	x, y := btcec.S256().ScalarMult(pub.X, pub.Y, priv.D.Bytes())
+	s.X = x
+	s.Y = y
+
+	h := sha256.Sum256(s.SerializeCompressed())
```

Comment:
Would the API be more flexible if we would return `s` so that the caller can also do something else with it than sha256?

Reply:
https://en.wikipedia.org/wiki/Elliptic-curve_Diffie%E2%80%93Hellman

Seems a bad idea